### PR TITLE
Improve typing for loops

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Part of #7037

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @justusschock @ananthsub @ninginthecloud

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 1]
```diff
@@ -51,11 +51,12 @@ def num_dataloaders(self) -> int:
     @property
     def dataloaders(self) -> Sequence[DataLoader]:
         """Returns the validation or test dataloaders."""
-        if self.trainer.testing:
-            return self.trainer.test_dataloaders
-        return self.trainer.val_dataloaders
+        dataloaders = self.trainer.test_dataloaders if self.trainer.testing else self.trainer.val_dataloaders
+        if dataloaders is None:
+            raise RuntimeError("Dataloaders should be available.")
```
Comment (rohitgr7) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:56]: any case where this error will be raised here?

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:56]: This is just to please mypy, I could've used an assertion too, but since this is public and loops are public, I chose this option

Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:56]: Should this be a ValueError ?

Comment (rohitgr7) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:56]: yeah, I think assertion would be enough. That's what we have at other places.

Comment (justusschock) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:56]: I prefer this one. It is cleaner. Ideally assertions only go into tests

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/optimization/optimizer_loop.py [hunk 2]
```diff
@@ -204,7 +204,9 @@ def on_run_start(  # type: ignore[override]
         self, batch: Any, optimizers: List[Tuple[int, Optimizer]], batch_idx: int
     ) -> None:
         self._batch_idx = batch_idx
-        self._indices, self._optimizers = zip(*optimizers)
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/loops/optimization/optimizer_loop.py:207]: ah wait. I removed exactly for that reason. mypy does correctly say a tuple is returneed

`self._indices` and `self._optimizers` are annotated as `List[...]`.

Was the intention that they are tuples? Should I change the List annotations instead?

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/optimization/optimizer_loop.py:207]: This is how it would work:
```py
>>> x = [(1, "a"), (2, "b")]
>>> zip(*x)
Out[3]: <zip at 0x7fdc6bf36d00>
>>> k, l = zip(*x)
>>> k
>>> Out[5]: (1, 2)
>>> l
>>> Out[6]: ('a', 'b')
```
And it seems the output is tuple. So I guess that's why it didn't match the typing. We could change the annotation of optimizers since it is used internally only.

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/optimization/optimizer_loop.py:207]: Yes, the code was cleaner.

**Review by tchaton — APPROVED** [review]
LGTM !
